torch/csrc/autograd/python_saved_variable_hooks.h

```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include <pybind11/pybind11.h>
+#include <torch/csrc/autograd/python_variable.h>
+#include <torch/csrc/autograd/saved_variable_hooks.h>
+#include <torch/csrc/python_headers.h>
+#include <ATen/ATen.h>
+
+namespace py = pybind11;
+
+namespace torch { namespace autograd {
+
+struct TORCH_API PySavedVariableHooks : public SavedVariableHooks {
+    PySavedVariableHooks(py::function &pack_hook, py::function &unpack_hook) : pack_hook_(pack_hook), unpack_hook_(unpack_hook){}
+
+    PyObject* call_pack_hook(at::Tensor &tensor) override {
+      return pack_hook_(py::reinterpret_steal<py::object>(THPVariable_Wrap(tensor))).release().ptr();
```

Comment:
note that this one liner makes it tricky to know when refcount bumps happen for the python objects.
You might want to split this into 3 lines to make it clear that you first get a THPVariable for which you own a reference, then you pass that to the hook function which returns another object with a reference. You then release the new object on return and the THPVariable refcount is decrefed on exit by the py::object destruction.

Reply:
Good point. Splitting in 3 lines made me realize that the `call_unpack_hook` is actually incorrect in its current form